### PR TITLE
178899484_add_websockets_to_solana_rpc_ruby

### PR DESCRIPTION
**Pull request by radek-polcode**
### Summary

Add websockets feature to gem.

1. Run `ruby demo_ws.rb` (uncomment methods you need, check the demo file for more info). 

Check README for more information.

### Other Information

If there's anything else that's important and relevant to your pull
request, mention that information here.

### CODE REVIEW

File: lib/solana_rpc_ruby/redis_client.rb [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+      host: 'localhost', 
+      expire: true, 
+      expiration_time: DEFAULT_EXPIRATION_TIME,
+      set_name: 'solana_sorted_set'
+    )
+      @redis = Redis.new(host: host)
+      @expire = expire
+      @expiration_time = expiration_time
+      @set_name = set_name
+      @score = get_highest_score.first[1]
```
Comment (pawjer) [inline lib/solana_rpc_ruby/redis_client.rb:17]: Use safe operator here. It caused some errors on starting the 'demo_ws' and 'demo_redis' scripts.

**Review by pawjer — APPROVED** [review]

File: lib/solana_rpc_ruby/helper_methods.rb [hunk 2]
```diff
@@ -12,5 +12,18 @@ def blank?(object)
...
+    # 
+    # @param method [String
+    # 
+    # @return [String]
+    def create_method_name(method)
+      return '' unless method && (method.is_a?(String) || method.is_a?(Symbol))
+
+      method.to_s.split('_').map.with_index do |string, i|
+        i == 0 ? string : string.capitalize
+      end.join
```
Comment (maciejkocylapc) [inline lib/solana_rpc_ruby/helper_methods.rb:26]: why not just use .camelize(:lower)

File: spec/lib/solana_rpc_ruby/websocket_client_spec.rb [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    let(:ws_testnet_cluster) { 'ws://api.testnet.solana.com' }
+
+    it 'uses ws cluster from config' do
+      api_client = described_class.new
+      
+      expect(api_client.cluster).to eq(SolanaRpcRuby.ws_cluster)
+    end
+
+    it 'fails without ws cluster passed in or set in config' do
+      SolanaRpcRuby.config do |c|
```
Comment (maciejkocylapc) [inline spec/lib/solana_rpc_ruby/websocket_client_spec.rb:15]: use more descriptive variables than 'c'

File: lib/solana_rpc_ruby/redis_client.rb (hunk 1 above)
Comment (radek-polcode) [reply lib/solana_rpc_ruby/redis_client.rb]: Redis is no longer used now.

File: lib/solana_rpc_ruby/helper_methods.rb (hunk 2 above)
Comment (radek-polcode) [reply lib/solana_rpc_ruby/helper_methods.rb:26]: It's a method from rails.

File: demo_ws.rb [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+# ws_method_wrapper.account_subscribe(account_id, &block)
+# ws_method_wrapper.logs_subscribe('all')
+# ws_method_wrapper.logs_subscribe('all', &block)
+# ws_method_wrapper.program_subscribe(program_id)
+# ws_method_wrapper.program_subscribe(program_id, &block)
+# ws_method_wrapper.root_subscribe
+ws_method_wrapper.root_subscribe(&block)
+# ws_method_wrapper.signature_subscribe('provide_signature')
+# ws_method_wrapper.slot_subscribe(&block)
+# ws_method_wrapper.slot_updates_subscribe
```
Comment (maciejkocylapc) [inline demo_ws.rb:48]: returns `{"jsonrpc":"2.0","error":{"code":-32601,"message":"Method not found"},"id":69018}`

**Comment by maciejkocylapc** [comment]
after subscription to websocket ends program should continue to work - currently it just stops (thread with subscription should end)

File: demo_ws.rb (hunk 4 above)
Comment (radek-polcode) [reply demo_ws.rb:48]: Method name should be `slots_updates_subscribe`, fixed, thanks.

**Pull request merged**